streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -655,10 +656,11 @@ public synchronized void onChange(final Thread thread,
                     // global stream thread has different invariants
                     final GlobalStreamThread.State newState = (GlobalStreamThread.State) abstractNewState;
                     globalThreadState = newState;
+                    final GlobalStreamThread.State oldState = (GlobalStreamThread.State) abstractOldState;
 
                     if (newState == GlobalStreamThread.State.RUNNING) {
                         maybeSetRunning();
-                    } else if (newState == GlobalStreamThread.State.DEAD) {
+                    } else if (newState == GlobalStreamThread.State.DEAD && oldState != GlobalStreamThread.State.PENDING_SHUTDOWN) {
                         log.error("Global thread has died. The streams application or client will now close to ERROR.");
                         closeToError();
```

Comment:
The global stream thread should already call `closeToError()` when the Streams uncaught exception handler is called in `run()` in `GlobalStreamThread`, shouldn't it?

Reply:
Why? The uncaught exception handler is user code and cannot set the client state;`closeToError()` is an internal method the user cannot call.

But thinking about it, it seem we are missing a call to `closeToError()` with `GlobalStreamThread#run()` exception handling blocks?

Reply:
> Why? The uncaught exception handler is user code and cannot set the client state;closeToError() is an internal method the user cannot call.

While it is correct that the uncaught exception is user code, the return value of the user code in the uncaught exception handler is evaluated in [`handleStreamsUncaughtException()`](https://github.com/apache/kafka/blob/c1071327c50613ae8647f1fa98f00f78a0cb5d76/streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java#L516). There `closeToError()` is called.